proxy/http/HttpTransactHeaders.cc

```diff
@@ -996,6 +1015,178 @@ HttpTransactHeaders::add_global_user_agent_header_to_request(OverridableHttpConf
 }
 
 void
+HttpTransactHeaders::add_forwarded_field_to_request(HttpTransact::State *s, HTTPHdr *request)
+{
+  namespace HF = HttpForwarded;
+
+  HF::OptionBitSet optSet = s->txn_conf->insert_forwarded;
+
+  if (optSet.any()) { // One or more Forwarded parameters enabled, so insert/append to Forwarded header.
+
+    ts::LocalBufferWriter<1024> hdr;
+
+    if (optSet[HF::Option::For] and ats_is_ip(&s->client_info.src_addr.sa)) {
+      // NOTE:  The logic within this if statement assumes that hdr is empty at this point.
+
+      hdr << "for=";
+
+      bool is_ipv6 = ats_is_ip6(&s->client_info.src_addr.sa);
+
+      if (is_ipv6) {
+        hdr << "\"[";
+      }
+
+      if (ats_ip_ntop(&s->client_info.src_addr.sa, hdr.auxBuffer(), hdr.remaining()) == nullptr) {
+        Debug("http_trans", "[add_forwarded_field_to_outgoing_request] ats_ip_ntop() call failed");
+        return;
+      }
+
+      // Fail-safe.
+      hdr.auxBuffer()[hdr.remaining() - 1] = '\0';
+
+      hdr.write(strlen(hdr.auxBuffer()));
```

Comment:
Does this copy the data on to itself?

Reply:
No, logically it appends the auxiliary buffer to the main buffer.  But, since the auxiliary buffer is just the space after the data in the main buffer in FixedBufferWriter, no actually copying is need, the index to the end of the main buffer data is just increased.

Reply:
No wait, you're right, this is wrong.

Reply:
Ah, there's an overload of `write`.

Reply:
It seems wrong to have to call `strlen` here. Surely the amount of data written is known. That would be one reason to use a nested `BufferWriter` on the `auxBuffer`. It might be a good API change to have `auxBuffer` return an `auxWriter` that's constructed to use that space.

Reply:
I don't think there is any way to avoid the call to strlen().  ats_ip_ntop() returns no info about how much data it put into the destination buffer.  On success, it just returns the destination buffer, rather than a pointer to the next empty byte in the destination buffer.

Reply:
Yeah, I think you're correct.